File: Daily/650.2-keys-keyboard.cpp
/*
 * @lc app=leetcode id=650 lang=cpp
 *
 * [650] 2 Keys Keyboard
 */

#include <algorithm>
#include <climits>
#include <vector>

// @lc code=start
class Solution {
 public:
  int minSteps(int n) {
    // dp[i] - The minimum steps to reach length i
    // dp[i] = dp[j] + 1 + (i - j) / j if i % j == 0 for j from 1 to i
    // Copy: +1
    // Pastes: (i-j)/j
    std::vector<int> dp(n + 1, INT_MAX);
    dp[1] = 0;

    for (int i = 2; i <= n; i++) {
      for (int j = i - 1; j >= 1; j--) {
        if (i % j == 0) {
          dp[i] = std::min(dp[i], /*dp[j] + 1 + (i - j) / j*/ dp[j] + i / j);
        }
      }
    }

    return dp[n];
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.minSteps(20);
}

File: src/2022-intern/650.2-keys-keyboard.js
/*
 * @lc app=leetcode id=650 lang=javascript
 *
 * [650] 2 Keys Keyboard
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var minSteps = function (n) {
  /**
   * 状态
   * 已经输入了多少？
   * 剪切板里面字符的数量？
   *
   * 转移
   * 复制
   * 粘贴
   */

  /**
   * dp[i][j] 当已经输入了 i 个字符，剪切板里面字符数量为 j 的时候，最小的操作次数
   *
   * base case
   * dp[n][?] = 0
   *
   * dp[i][j] =
   * if j === 0
   *  1+dp[i][i]
   * else if(j<i)
   *  1+min(
   *  dp[i][i],  // copy
   *  i+j>n?Infinity: dp[i+j][j]
   *  )
   * else
   *  1+(i+j>n?Infinity:dp[i+j][j])
   *
   * 返回 dp[1][0]
   */

  /**
     * 优化
     * dp[i] 输入 i 个字符所需的最小操作次数
     * 
     * base case
     * dp[1] = 0 
     * 
     * k from 1 to i-1
     * if (i % k === 0)
         min = Math.min(min,
                dp[k]   // 得到被复制部分
                + 1     // 复制
                + (i / k - 1) // 粘贴这么多次
            );
     * 
     * 返回 dp[n]
     */

  const dp = new Array(n + 1);
  dp[1] = 0;

  for (let i = 2; i <= n; i++) {
    let min = Infinity;
    for (let k = 1; k <= i - 1; k++) {
      if (i % k === 0) {
        min = Math.min(min, dp[k] + i / k);
      }
    }
    dp[i] = min;
  }

  return dp[n];
};
// @lc code=end
